chain/client/src/chunk_executor_actor.rs

```diff
@@ -0,0 +1,516 @@
+        shard_id: ShardId,
+        mode: ApplyChunksMode,
+    ) -> Result<ShardContext, Error> {
+        let cares_about_shard_this_epoch =
+            self.shard_tracker.cares_about_shard(me, prev_hash, shard_id, true);
+        let cares_about_shard_next_epoch =
+            self.shard_tracker.will_care_about_shard(me, prev_hash, shard_id, true);
+        let cared_about_shard_prev_epoch =
+            self.shard_tracker.cared_about_shard_in_prev_epoch(me, prev_hash, shard_id, true);
+        let should_apply_chunk = get_should_apply_chunk(
+            mode,
+            cares_about_shard_this_epoch,
+            cares_about_shard_next_epoch,
+            cared_about_shard_prev_epoch,
+        );
+        let shard_uid = shard_id_to_uid(self.epoch_manager.as_ref(), shard_id, epoch_id)?;
+        Ok(ShardContext { shard_uid, should_apply_chunk })
+    }
+
+    fn chain_update(&mut self) -> ChainUpdate {
+        ChainUpdate::new(
+            &mut self.chain_store,
+            self.epoch_manager.clone(),
+            self.runtime_adapter.clone(),
+            // Since we don't produce blocks, this argument is irrelevant.
+            DoomslugThresholdMode::NoApprovals,
+        )
+    }
+
+    /// Returns keys from block_receipts_cache for prev_block_hash that correspond to all incoming
+    /// receipts that we care about grouped by destination shard id.
+    /// Returns None if some of the receipts are still missing.
+    fn all_incoming_receipts_keys(
+        &mut self,
+        me: Option<&AccountId>,
+        prev_block_hash: &CryptoHash,
+        block_hash: &CryptoHash,
+    ) -> Result<Option<Vec<(ShardId, ShardId)>>, Error> {
+        let Some(block_receipts) = self.block_receipts_cache.get(prev_block_hash) else {
+            return Ok(None);
+        };
+        let epoch_id = self.epoch_manager.get_epoch_id(prev_block_hash)?;
+        let shard_ids = self.epoch_manager.shard_ids(&epoch_id)?;
+
+        let mut keys = Vec::<(ShardId, ShardId)>::new();
+        for to_shard_id in &shard_ids {
+            let is_me = true;
+            // We cannot filter out correctly when receiving receipts since we may receive them
+            // before we know about the corresponding block and can decide which receipts we care
+            // about.
+            if !self.shard_tracker.cares_about_shard_this_or_next_epoch(
+                me,
+                &block_hash,
+                *to_shard_id,
+                is_me,
+            ) {
+                continue;
+            }
+            for from_shard_id in &shard_ids {
+                if !block_receipts.contains_key(&(*from_shard_id, *to_shard_id)) {
+                    return Ok(None);
+                }
+                keys.push((*from_shard_id, *to_shard_id));
+            }
+        }
+        Ok(Some(keys))
+    }
+
+    #[instrument(target = "chunk_executor", level = "debug", skip_all, fields(%prev_block_hash, %block_hash, ?me))]
+    fn try_save_incoming_receipts(
```

Comment:
here let's rename `prev_block_hash` -> `block_hash` and try removing dependency on the next block hash